backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/predicate/impl/ElasticsearchTermsPredicate.java

```diff
@@ -0,0 +1,177 @@
+import org.hibernate.search.backend.elasticsearch.search.impl.ElasticsearchSearchContext;
+import org.hibernate.search.backend.elasticsearch.search.impl.ElasticsearchSearchValueFieldContext;
+import org.hibernate.search.backend.elasticsearch.search.predicate.impl.AbstractElasticsearchSingleFieldPredicate;
+import org.hibernate.search.backend.elasticsearch.search.predicate.impl.PredicateRequestContext;
+import org.hibernate.search.backend.elasticsearch.types.codec.impl.ElasticsearchFieldCodec;
+import org.hibernate.search.engine.backend.types.converter.spi.DslConverter;
+import org.hibernate.search.engine.reporting.spi.EventContexts;
+import org.hibernate.search.engine.search.common.ValueConvert;
+import org.hibernate.search.engine.search.predicate.SearchPredicate;
+import org.hibernate.search.engine.search.predicate.spi.TermsPredicateBuilder;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+public class ElasticsearchTermsPredicate extends AbstractElasticsearchSingleFieldPredicate {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private static final JsonObjectAccessor TERMS_ACCESSOR = JsonAccessor.root().property( "terms" ).asObject();
+	private static final JsonObjectAccessor TERM_ACCESSOR = JsonAccessor.root().property( "term" ).asObject();
+
+	// for boolean query:
+	private static final JsonObjectAccessor BOOL_ACCESSOR = JsonAccessor.root().property( "bool" ).asObject();
+	private static final JsonArrayAccessor MUST_ACCESSOR = JsonAccessor.root().property( "must" ).asArray();
+
+	private static final JsonAccessor<JsonElement> VALUE_ACCESSOR = JsonAccessor.root().property( "value" );
+
+	private final JsonElement term;
+	private final JsonElement[] terms;
+	private boolean allMatch;
+
+	public ElasticsearchTermsPredicate(Builder builder) {
+		super( builder );
+		this.term = builder.term;
+		this.terms = builder.terms;
+		this.allMatch = builder.allMatch;
+	}
+
+	@Override
+	protected JsonObject doToJsonQuery(PredicateRequestContext context, JsonObject outerObject,
+			JsonObject innerObject) {
+
+		// single term
+		if ( term != null ) {
+			VALUE_ACCESSOR.set( innerObject, term );
+			JsonObject middleObject = new JsonObject();
+			middleObject.add( absoluteFieldPath, innerObject );
+			TERM_ACCESSOR.set( outerObject, middleObject );
+			return outerObject;
+		}
+
+		// multiple terms in OR
+		if ( !allMatch ) {
+			JsonArray jsonArray = new JsonArray( terms.length );
+			for ( JsonElement element : terms ) {
+				jsonArray.add( element );
+			}
+
+			innerObject.add( absoluteFieldPath, jsonArray );
+			TERMS_ACCESSOR.set( outerObject, innerObject );
+			return outerObject;
+		}
+
+		// multiple terms in AND
+		// using a boolean query here:
```

Comment:
Can't you just use `minimum_should_match_field`? That seems more efficient, especially if there are many terms.

See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-terms-set-query.html#terms-set-top-level-params

In fact, if there is a large number of terms, a boolean query simply won't work: you can only stuff a certain number of clauses into a boolean query.

Could you add a test with many terms? By "many" I mean more than what the maximum number of clauses allowed in a boolean query (`BooleanQuery.maxClauseCount`); that's definitely where this predicate will shine.

Reply:
Reading the documentation, it seems that there isn't a `minimum_should_match` option here. From my understanding `minimum_should_match_field` needs a field on the index to store the `minimum_should_match` value.
There is `minimum_should_match_script` but a script probably needs to be executed somehow, so I'm not sure that would be a great optimization.

Reply:
Right, I misread that part...

Supposedly `painless` scripts are pre-compiled before running the query... So I guess it could perform well?

Anyway, you're right, it seems doubtful this optimization would have a great impact. And considering we have that exact same limitation of accepting a limited number of terms for all `matchingAll` implementations (text and numbers, Lucene and Elasticsearch), I guess it doesn't hurt to keep that limitation here.

Reply:
We'll need to document the limitation somewhere, though ;)